File: src/csharp/Grpc.Core/Interceptors/Interceptor.cs
#region Copyright notice and license

// Copyright 2017 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Reflection;
using System.Threading.Tasks;
using Grpc.Core.Internal;

namespace Grpc.Core.Interceptors
{
    /// <summary>
    /// Carries along the context associated with intercepted invocations on the client side.
    /// This is an EXPERIMENTAL API.
    /// </summary>
    public class ClientInterceptorContext<TRequest, TResponse>
        where TRequest : class
        where TResponse : class
    {
        /// <summary>
        /// Creates a new instance of <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}" />
        /// with the specified method, host, and call options.
        /// </summary>
        /// <param name="method">A <see cref="Grpc.Core.Method{TRequest, TResponse}"/> object representing the method to be invoked.</param>
        /// <param name="host">The host to dispatch the current call to.</param>
        /// <param name="options">A <see cref="Grpc.Core.CallOptions"/> instance containing the call options of the current call.</param>

        public ClientInterceptorContext(Method<TRequest, TResponse> method, string host, CallOptions options)
        {
            Method = method;
            Host = host;
            Options = options;
        }

        /// <summary>
        /// Gets the <see cref="Grpc.Core.Method{TRequest, TResponse}"/> representing
        /// the method to be invoked.
        /// </summary>
        public Method<TRequest, TResponse> Method { get; }

        /// <summary>
        /// Gets the host that the currect invocation will be dispatched to.
        /// </summary>

        public string Host { get; }

        /// <summary>
        /// Gets the <see cref="Grpc.Core.CallOptions"/> structure representing the
        /// call options associated with the current invocation.
        /// </summary>

        public CallOptions Options { get; }
    }

    /// <summary>
    /// Serves as the base class for gRPC interceptors.
    /// This is an EXPERIMENTAL API.
    /// </summary>
    public abstract class Interceptor
    {
        /// <summary>
        /// Represents a continuation for intercepting simple blocking invocations.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this invocation.</typeparam>
        /// <typeparam name="TResponse">Response message type for this invocation.</typeparam>
        /// <param name="request">The request value to continue the invocation with.</param>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// instance to pass to the next step in the invocation process.
        /// </param>
        public delegate TResponse BlockingUnaryCallContinuation<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Represents a continuation for intercepting simple asynchronous invocations.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this invocation.</typeparam>
        /// <typeparam name="TResponse">Response message type for this invocation.</typeparam>
        /// <param name="request">The request value to continue the invocation with.</param>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// instance to pass to the next step in the invocation process.
        /// </param>
        public delegate AsyncUnaryCall<TResponse> AsyncUnaryCallContinuation<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Represents a continuation for intercepting asynchronous server-streaming invocations.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this invocation.</typeparam>
        /// <typeparam name="TResponse">Response message type for this invocation.</typeparam>
        /// <param name="request">The request value to continue the invocation with.</param>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// instance to pass to the next step in the invocation process.
        /// </param>
        public delegate AsyncServerStreamingCall<TResponse> AsyncServerStreamingCallContinuation<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Represents a continuation for intercepting asynchronous client-streaming invocations.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this invocation.</typeparam>
        /// <typeparam name="TResponse">Response message type for this invocation.</typeparam>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// instance to pass to the next step in the invocation process.
        /// </param>
        public delegate AsyncClientStreamingCall<TRequest, TResponse> AsyncClientStreamingCallContinuation<TRequest, TResponse>(ClientInterceptorContext<TRequest, TResponse> context)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Represents a continuation for intercepting asynchronous duplex invocations.
        /// </summary>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// instance to pass to the next step in the invocation process.
        /// </param>
        public delegate AsyncDuplexStreamingCall<TRequest, TResponse> AsyncDuplexStreamingCallContinuation<TRequest, TResponse>(ClientInterceptorContext<TRequest, TResponse> context)
            where TRequest : class
            where TResponse : class;

        /// <summary>
        /// Intercepts a blocking invocation of a simple remote call.
        /// </summary>
        /// <param name="request">The request message of the invocation.</param>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// associated with the current invocation.
        /// </param>
        /// <param name="continuation">
        /// The callback that continues the invocation process.
        /// This can be invoked zero or more times by the interceptor.
        /// </param>
        /// <returns>The response messaage of the current invocation.</returns>
        public virtual TResponse BlockingUnaryCall<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context, BlockingUnaryCallContinuation<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(request, context);
        }

        /// <summary>
        /// Intercepts an asynchronous invocation of a simple remote call.
        /// </summary>
        /// <param name="request">The request message of the invocation.</param>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// associated with the current invocation.
        /// </param>
        /// <param name="continuation">
        /// The callback that continues the invocation process.
        /// This can be invoked zero or more times by the interceptor.
        /// </param>
        public virtual AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context, AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(request, context);
        }

        /// <summary>
        /// Intercepts an asynchronous invocation of a streaming remote call.
        /// </summary>
        /// <param name="request">The request message of the invocation.</param>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// associated with the current invocation.
        /// </param>
        /// <param name="continuation">
        /// The callback that continues the invocation process.
        /// This can be invoked zero or more times by the interceptor.
        /// </param>
        public virtual AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(TRequest request, ClientInterceptorContext<TRequest, TResponse> context, AsyncServerStreamingCallContinuation<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(request, context);
        }

        /// <summary>
        /// Intercepts an asynchronous invocation of a client streaming call.
        /// </summary>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// associated with the current invocation.
        /// </param>
        /// <param name="continuation">
        /// The callback that continues the invocation process.
        /// This can be invoked zero or more times by the interceptor.
        /// </param>
        public virtual AsyncClientStreamingCall<TRequest, TResponse> AsyncClientStreamingCall<TRequest, TResponse>(ClientInterceptorContext<TRequest, TResponse> context, AsyncClientStreamingCallContinuation<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(context);
        }

        /// <summary>
        /// Intercepts an asynchronous invocation of a duplex streaming call.
        /// </summary>
        /// <param name="context">
        /// The <see cref="Grpc.Core.Interceptors.ClientInterceptorContext{TRequest, TResponse}"/>
        /// associated with the current invocation.
        /// </param>
        /// <param name="continuation">
        /// The callback that continues the invocation process.
        /// This can be invoked zero or more times by the interceptor.
        /// </param>
        public virtual AsyncDuplexStreamingCall<TRequest, TResponse> AsyncDuplexStreamingCall<TRequest, TResponse>(ClientInterceptorContext<TRequest, TResponse> context, AsyncDuplexStreamingCallContinuation<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(context);
        }

        /// <summary>
        /// Server-side handler for intercepting unary calls.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this method.</typeparam>
        /// <typeparam name="TResponse">Response message type for this method.</typeparam>
        public virtual Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
            TRequest request,
            ServerCallContext context,
            UnaryServerMethod<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(request, context);
        }

        /// <summary>
        /// Server-side handler for intercepting client streaming call.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this method.</typeparam>
        /// <typeparam name="TResponse">Response message type for this method.</typeparam>
        public virtual Task<TResponse> ClientStreamingServerHandler<TRequest, TResponse>(
            IAsyncStreamReader<TRequest> requestStream,
            ServerCallContext context,
            ClientStreamingServerMethod<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(requestStream, context);
        }

        /// <summary>
        /// Server-side handler for intercepting server streaming calls.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this method.</typeparam>
        /// <typeparam name="TResponse">Response message type for this method.</typeparam>
        public virtual Task ServerStreamingServerHandler<TRequest, TResponse>(
            TRequest request,
            IServerStreamWriter<TResponse> responseStream,
            ServerCallContext context,
            ServerStreamingServerMethod<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(request, responseStream, context);
        }

        /// <summary>
        /// Server-side handler for intercepting bidi streaming calls.
        /// </summary>
        /// <typeparam name="TRequest">Request message type for this method.</typeparam>
        /// <typeparam name="TResponse">Response message type for this method.</typeparam>
        public virtual Task DuplexStreamingServerHandler<TRequest, TResponse>(
            IAsyncStreamReader<TRequest> requestStream,
            IServerStreamWriter<TResponse> responseStream,
            ServerCallContext context,
            DuplexStreamingServerMethod<TRequest, TResponse> continuation)
            where TRequest : class
            where TResponse : class
        {
            return continuation(requestStream, responseStream, context);
        }

        /// <summary>
        /// Used by <c>WrapDelegate</c> function to wire up a non-generic
        /// handler to a type-safe generic interceptor function with the
        /// correct type arguments.
        /// </summary>
        private static class WrapUtil<TRequest, TResponse>
            where TRequest : class
            where TResponse : class
        {
            public static UnaryServerMethod<TRequest, TResponse> Unary(
                UnaryServerMethod<TRequest, TResponse> handler,
                Interceptor interceptor)
            {
                return (request, context) =>
                    interceptor.UnaryServerHandler<TRequest, TResponse>(request, context, handler);
            }

            public static ClientStreamingServerMethod<TRequest, TResponse> ClientStreaming(
                ClientStreamingServerMethod<TRequest, TResponse> handler,
                Interceptor interceptor)
            {
                return (request, context) =>
                    interceptor.ClientStreamingServerHandler<TRequest, TResponse>(request, context, handler);
            }

            public static ServerStreamingServerMethod<TRequest, TResponse> ServerStreaming(
                ServerStreamingServerMethod<TRequest, TResponse> handler,
                Interceptor interceptor)
            {
                return (request, response, context) =>
                    interceptor.ServerStreamingServerHandler<TRequest, TResponse>(request, response, context, handler);
            }

            public static DuplexStreamingServerMethod<TRequest, TResponse> DuplexStreaming(
                DuplexStreamingServerMethod<TRequest, TResponse> handler,
                Interceptor interceptor)
            {
                return (request, response, context) =>
                    interceptor.DuplexStreamingServerHandler<TRequest, TResponse>(request, response, context, handler);
            }
        }

        /// <summary>
        /// Returns a wrapped delegate that intercepts the calls to the
        /// given delegate <c>d</c> in a type-safe fashion.
        /// This is necessary because the interceptor does not have
        /// a priori information about the generic type of each handler
        /// and needs to reconsturct that information by reflecting over
        /// the delegate and matching it with the appropriate generic
        /// interceptor function.
        /// </summary>
        private Delegate WrapDelegate(Delegate d)
        {
            if (d == null)
            {
                return d;
            }

            var dType = d.GetType().GetTypeInfo();
            if (!dType.IsGenericType)
            {
                return d;
            }

            var genericType = dType.GetGenericTypeDefinition();
            if (genericType == typeof(UnaryServerMethod<,>))
            {
                return (Delegate)typeof(WrapUtil<,>).GetTypeInfo()
                    .MakeGenericType(dType.GetGenericArguments()).GetTypeInfo()
                    .GetMethod("Unary").Invoke(null, new object[] { d, this });
            }
            else if (genericType == typeof(ClientStreamingServerMethod<,>))
            {
                return (Delegate)typeof(WrapUtil<,>).GetTypeInfo()
                    .MakeGenericType(dType.GetGenericArguments()).GetTypeInfo()
                    .GetMethod("ClientStreaming").Invoke(null, new object[] { d, this });
            }
            else if (genericType == typeof(ServerStreamingServerMethod<,>))
            {
                return (Delegate)typeof(WrapUtil<,>).GetTypeInfo()
                    .MakeGenericType(dType.GetGenericArguments()).GetTypeInfo()
                    .GetMethod("ServerStreaming").Invoke(null, new object[] { d, this });
            }
            else if (genericType == typeof(DuplexStreamingServerMethod<,>))
            {
                return (Delegate)typeof(WrapUtil<,>).GetTypeInfo()
                    .MakeGenericType(dType.GetGenericArguments()).GetTypeInfo()
                    .GetMethod("DuplexStreaming").Invoke(null, new object[] { d, this });
            }

            return d;
        }

        /// <summary>
        /// Decorates an instance of <see cref="Grpc.Core.Internal.IServerCallHandler"/>
        /// and intercepts the execution its handlers.
        /// </summary>
        /// <returns>
        /// Returns a decorated <see cref="Grpc.Core.Internal.IServerCallHandler"/>
        /// if <c>handler</c> implements <see cref="Grpc.Core.Internal.IInterceptableCallHandler" />,
        /// unmodified <c>handler</c> otherwise.
        /// </returns>
        internal IServerCallHandler WrapServerCallHandler(IServerCallHandler handler)
        {
            var interceptable = handler as IInterceptableCallHandler;
            if (interceptable == null)
            {
                return handler;
            }

            return interceptable.Intercept(WrapDelegate);
        }
    }
}

File: src/python/grpcio/grpc/_interceptor.py
# Copyright 2017 gRPC authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Implementation of gRPC Python interceptors."""

import collections

import grpc


class _ServicePipeline(object):

    def __init__(self, interceptors):
        self.interceptors = tuple(interceptors)

    def _continuation(self, thunk, index):
        return lambda context: self._intercept_at(thunk, index, context)

    def _intercept_at(self, thunk, index, context):
        if index < len(self.interceptors):
            interceptor = self.interceptors[index]
            thunk = self._continuation(thunk, index + 1)
            return interceptor.intercept_service(thunk, context)
        return thunk(context)

    def execute(self, thunk, context):
        return self._intercept_at(thunk, 0, context)


def service_pipeline(interceptors):
    if interceptors:
        return _ServicePipeline(interceptors)
    return None


class _ClientCallDetails(
        collections.namedtuple('_ClientCallDetails', (
            'method', 'request_streaming', 'response_streaming',
            'request_serializer', 'response_deserializer', 'timeout',
            'metadata', 'credentials'), grpc.ClientCallDetails)):
    pass


def _continuation(thunk):

    def continuation(client_call_details, request_iterator):
        if not client_call_details.request_streaming:
            next(request_iterator)
            return thunk(client_call_details)(
                request,
                timeout=client_call_details.timeout,
                metadata=client_call_details.metadata,
                credentials=client_call_details.credentials)


class _UnaryUnaryMultiCallable(grpc.UnaryUnaryMultiCallable):

    def __init__(self, thunk, interceptor, method, request_serializer,
                 response_deserializer):
        self._thunk = thunk
        self._interceptor = interceptor
        self._method = method
        self._request_serializer = request_serializer
        self._response_deserializer = response_deserializer

    def __call__(self, request, timeout=None, metadata=None, credentials=None):
        future = self.future(
            request,
            timeout=timeout,
            metadata=metadata,
            credentials=credentials)
        return future.result()

    def with_call(self, request, timeout=None, metadata=None, credentials=None):
        future = self.future(
            request,
            timeout=timeout,
            metadata=metadata,
            credentials=credentials)
        return future.result(), future

    def future(self, request, timeout=None, metadata=None, credentials=None):

        response_iterator = self._interceptor.intercept_call(
            _continuation(self._thunk),
            _ClientCallDetails(
                self._method, False, False, self._request_serializer,
                self._response_deserializer, timeout, metadata, credentials),
            (request,))


class _UnaryStreamMultiCallable(grpc.UnaryStreamMultiCallable):

    def __init__(self, thunk, interceptor, method, request_serializer,
                 response_deserializer):
        self._thunk = thunk
        self._interceptor = interceptor
        self._method = method
        self._request_serializer = request_serializer
        self._response_deserializer = response_deserializer

    def __call__(self, request, timeout=None, metadata=None, credentials=None):
        return self._interceptor.intercept_call(
            continuation,
            _ClientCallDetails(self._method, timeout, metadata, credentials),
            request)

    def _continuation(self, client_call_details, request):
        return self._thunk(client_call_details.method)(
            request,
            timeout=client_call_details.timeout,
            metadata=client_call_details.metadata,
            credentials=client_call_details.credentials)


class _StreamUnaryMultiCallable(grpc.StreamUnaryMultiCallable):

    def __init__(self, thunk, interceptor, method, request_serializer,
                 response_deserializer):
        self._thunk = thunk
        self._interceptor = interceptor
        self._method = method
        self._request_serializer = request_serializer
        self._response_deserializer = response_deserializer

    def __call__(self,
                 request_iterator,
                 timeout=None,
                 metadata=None,
                 credentials=None):
        future = self.future(
            request_iterator,
            timeout=timeout,
            metadata=metadata,
            credentials=credentials)
        return future.result()

    def with_call(self,
                  request_iterator,
                  timeout=None,
                  metadata=None,
                  credentials=None):
        future = self.future(
            request_iterator,
            timeout=timeout,
            metadata=metadata,
            credentials=credentials)
        return future.result(), future

    def future(self,
               request_iterator,
               timeout=None,
               metadata=None,
               credentials=None):

        def continuation(client_call_details, request_iterator):
            return self._thunk(client_call_details.method).future(
                request_iterator,
                timeout=client_call_details.timeout,
                metadata=client_call_details.metadata,
                credentials=client_call_details.credentials)

        return self._interceptor.intercept_stream_unary(
            continuation,
            _ClientCallDetails(self._method, timeout, metadata, credentials),
            request_iterator)


class _StreamStreamMultiCallable(grpc.StreamStreamMultiCallable):

    def __init__(self, thunk, interceptor, method, request_serializer,
                 response_deserializer):
        self._thunk = thunk
        self._interceptor = interceptor
        self._method = method
        self._request_serializer = request_serializer
        self._response_deserializer = response_deserializer

    def __call__(self,
                 request_iterator,
                 timeout=None,
                 metadata=None,
                 credentials=None):

        def continuation(client_call_details, request_iterator):
            return self._thunk(client_call_details.method)(
                request_iterator,
                timeout=client_call_details.timeout,
                metadata=client_call_details.metadata,
                credentials=client_call_details.credentials)

        return self._interceptor.intercept_call(
            continuation,
            _ClientCallDetails(self._method, timeout, metadata, credentials),
            request_iterator)


class _Channel(grpc.Channel):

    def __init__(self, channel, interceptor):
        self._channel = channel
        self._interceptor = interceptor

    def subscribe(self, *args, **kwargs):
        self._channel.subscribe(*args, **kwargs)

    def unsubscribe(self, *args, **kwargs):
        self._channel.unsubscribe(*args, **kwargs)

    def unary_unary(self,
                    method,
                    request_serializer=None,
                    response_deserializer=None):
        return _UnaryUnaryMultiCallable(self._thunk, self._interceptor, method,
                                        request_serializer,
                                        response_deserializer)

    def unary_stream(self,
                     method,
                     request_serializer=None,
                     response_deserializer=None):
        return _UnaryStreamMultiCallable(self._thunk, self._interceptor, method,
                                         request_serializer,
                                         response_deserializer)

    def stream_unary(self,
                     method,
                     request_serializer=None,
                     response_deserializer=None):
        return _StreamUnaryMultiCallable(self._thunk, self._interceptor, method,
                                         request_serializer,
                                         response_deserializer)

    def stream_stream(self,
                      method,
                      request_serializer=None,
                      response_deserializer=None):
        return _StreamStreamMultiCallable(self._thunk, self._interceptor,
                                          method, request_serializer,
                                          response_deserializer)

    def _thunk(self, call):
        if not call.request_streaming and not call.response_streaming:
            return self._channel.unary_unary(
                call.method,
                request_serializer=call.request_serializer,
                response_deserializer=call.response_deserializer)
        if not call.request_streaming and call.response_streaming:
            return self._channel.unary_stream(
                call.method,
                request_serializer=call.request_serializer,
                response_deserializer=call.response_deserializer)
        if call.request_streaming and not call.response_streaming:
            return self._channel.stream_unary(
                call.method,
                request_serializer=call.request_serializer,
                response_deserializer=call.response_deserializer)
        if call.request_streaming and call.response_streaming:
            return self._channel.stream_stream(
                call.method,
                request_serializer=call.request_serializer,
                response_deserializer=call.response_deserializer)
        raise ValueError('call is not a valid ClientCallDetails value')


def intercept_channel(channel, *interceptors):
    for interceptor in reversed(list(interceptors)):
        if not isinstance(interceptor, grpc.ClientInterceptor):
            raise TypeError('interceptor must be grpc.ClientInterceptor')
        channel = _Channel(channel, interceptor)
    return channel
